### PR TITLE
feat: add `depth` to objects

### PR DESCRIPTION
**Pull request by endbr64**
TODO:
- [x] go through the initial values and decide whether `depth` should start at 0 or 1 for aggregation (fold, etc.)
    - i don't think a depth of `0` ever makes sense
- [x] index task for object
- [x] tests

### CODE REVIEW

File: packages/server/src/database.rs [hunk 1]
```diff
@@ -159,7 +163,8 @@ async fn migration_0000(database: &Database) -> tg::Result<()> {
 				index_status text,
 				index_started_at text,
 				touched_at text,
-				weight integer
+				weight integer,
+				depth integer
```
Comment (nitsky) [inline packages/server/src/database.rs]: Please keep the columns sorted.

File: packages/cli/src/object/get.rs [hunk 2]
```diff
@@ -23,6 +23,12 @@ impl Cli {
 			eprint!("∅");
 		}
 		eprintln!();
+		if let Some(depth) = metadata.depth {
+			eprint!("{depth}");
```
Comment (nitsky) [inline packages/cli/src/object/get.rs:28]: You need to render a label here, see above.

Comment (endbr64) [reply packages/cli/src/object/get.rs:28]: Done.

File: packages/server/src/database.rs (hunk 1 above)
Comment (endbr64) [reply packages/server/src/database.rs]: Done.

File: packages/server/src/blob/create.rs [hunk 3]
```diff
@@ -93,13 +94,13 @@ impl Server {
...
-								values ({p}1, {p}2, {p}3, {p}4, {p}5, {p}6)
-								on conflict (id) do update set touched_at = {p}6;
+								insert into objects (id, bytes, complete, count, depth, weight, touched_at)
+								values ({p}1, {p}2, {p}3, {p}4, {p}5, {p}6, {p}7)
+								on conflict (id) do update set touched_at = {p}7;
 							"
 						);
 						let now = time::OffsetDateTime::now_utc().format(&Rfc3339).unwrap();
-						let params = db::params![&id, &data.bytes, 1, 1, size, now];
+						let params = db::params![&id, &data.bytes, 1, 1, 0, size, now];
```
Comment (nitsky) [inline packages/server/src/blob/create.rs]: Isn't the depth of a leaf supposed to be 1, not 0?

File: packages/server/src/object/index.rs [hunk 4]
```diff
@@ -260,6 +262,32 @@ impl Server {
...
+				// Set the depth.
+				let p = connection.p();
+				let statement = formatdoc!(
+					"
+						update objects
+						set depth = {p}1
+						where id = {p}2;
+					"
+				);
+				let params = db::params![depth.unwrap() + 1, id];
```
Comment (nitsky) [inline packages/server/src/object/index.rs]: Won't you hit this unwrap for objects with no children? I think you want `depth.map_or(1, |depth| depth + 1)`.

Comment (endbr64) [reply packages/server/src/object/index.rs]: I used `.filter_map` above (pushed after your comment), which I think should filter out objects with no children.

File: packages/server/src/blob/create.rs (hunk 3 above)
Comment (endbr64) [reply packages/server/src/blob/create.rs]: Fixed.

File: packages/server/src/object/index.rs (hunk 4 above)
Comment (endbr64) [reply packages/server/src/object/index.rs]: Changed it to use [`.try_fold()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.try_fold), which should short-circuit and set the result to `None` when the internal closure returns `None`.

**Pull request merged**